### PR TITLE
Update matlab interface

### PR DESCRIPTION
**Pull request by bryanwweber**
I tried to split the documentation updates and the bug fixes/improvements as best I could. I might still go back and rewrite the commits to better separate the changes, but it might not be worth it.

Bryan

### CODE REVIEW

File: doc/SConscript [hunk 1]
```diff
@@ -34,14 +34,170 @@ if localenv['sphinx_docs']:
...
+                docstring += '    ' + line.split('  ')[1] + '\n'
+            except IndexError:
+                docstring += line + '\n'
+            # Skip the next line, which is a duplicate of the first. It is here
+            # because MATLAB doesn't show the function definition in its help.
+            in_file.readline()
+
+            # For the rest of the lines in the file, get the line if it is
+            # in the first unbroken comment section and add it to the docstring.
+            for i, line in enumerate(in_file.readlines()):
```
Comment (speth) [inline doc/SConscript:63]: You can drop the `enumerate` here, since you aren't using `i`.

File: doc/SConscript [hunk 2]
```diff
@@ -34,14 +34,170 @@ if localenv['sphinx_docs']:
             build(b)
             localenv.Depends(sphinxdocs, b)
 
+    # Set up functions to pseudo-autodoc the MATLAB toolbox
+    def extract_matlab_docstring(mfile):
```
Comment (speth) [inline doc/SConscript:38]: I think I would move the definitions of all of these helper functions out of the `if localenv['sphinx_docs']:` block, mostly to save the extra level of indentation.

**Comment by bryanwweber** [comment]
@speth Thanks for the notes so far. I think it would probably be best to create a new branch and cherry pick the changes from this branch so I can get a clean history. I'll split the addition of the documentation to the m-files, the changes to SConscript, and the changes to the m-file code into separate commits. The documentation thus won't necessarily match the actual code, but it will only be at one commit that it doesn't match, and since I'd be making all the commits at the same time, the head will always be up to date (and anyways, the documentation doesn't match what's there right now, so...).

File: doc/sphinx/matlab/code-docs/matlab-template.rst.in [hunk 3]
```diff
@@ -0,0 +1,4 @@
+
+@title@
+
+@matlab_docstrings@
```
Comment (bryanwweber) [inline doc/sphinx/matlab/code-docs/matlab-template.rst.in:4]: Need to add a newline at the end of this file.

**Comment by speth** [comment]
@bryanwweber You can also just use `git rebase -i master` on this branch to rework the set of commits here, rather than cherry-picking onto a different branch, but you can do whichever way makes the most sense to you. Splitting off the changes to the SConscript and other Sphinx-related changes would probably be useful. I agree that the commits with the big doc changes should be separate from any actual code changes, so the latter stand out more clearly.

File: doc/SConscript [hunk 4]
```diff
@@ -34,14 +34,170 @@ if localenv['sphinx_docs']:
...
+            return sig
+
+    # This dictionary lists every class in the toolbox, plus all the files in
+    # the root toolbox directory. The keys of the dictionary are the file names
+    # to which the documentation will be written. The value associated with each
+    # key is a list. The first element of the list is the header of the page. The
+    # other elements are lists of lists. The first of these nested lists is the
+    # class folder, the second is the files associated with a class in the root
+    # toolbox folder. Either of these lists may be empty.
+    files = {
```
Comment (speth) [inline doc/SConscript:103]: I think this data structure could be improved on a bit. Here's my suggestion:

``` python
from collections import namedtuple
Page = namedtuple('Page', 'name title objects')

pages = [Page('importing', 'Importing Phase Objects', 
         ['@Solution', '@Mixture']),
         Page('thermodynamics', 'Thermodynamic Properties', ['@ThermoPhase']),
         Page('kinetics', 'Chemical Kinetics', ['@Kinetics']),
         Page('transport', 'Transport Properties', ['@Transport'])
         Page('zero-dim', 'Zero-Dimensional Reactor Networks', 
              ['@Func', '@Reactor', '@ReactorNet', '@FlowDevice', '@Wall']),
         Page('one-dim': 'One-Dimensional Reacting Flows', 
              ['1D/@Domain1D', '1D/@Stack'])
         Page('data', 'Built-In Thermochemical Data',
              ['air.m', 'constants.m', 'gasconstant.m', 'GRI30.m', 'Hydrogen.m',
               'Methane.m', 'Nitrogen.m', 'oneatm.m', 'Oxygen.m', 'Water.m']),
         Page('utilities', 'Utility Functions',
              ['adddir.m', 'ck2cti.m', 'cleanup.m', 'geterr.m', '@XML_Node']),
         Page('interface', 'Interfaces', ['@Interface'])
         ]

extra = {'@Solution': ('IdealGasMix.m', 'importPhase.m'),
         '@Func': ('gaussian.m', 'polynom.m'),
         '@Reactor': ('conhp.m', 'ConstPressureReactor.m', 'conuv.m',
                      'FlowReactor.m', 'IdealGasConstPressureReactor.m',
                      'IdealGasReactor.m', 'reactor_ode.m', 'Reservoir.m'),
         '@FlowDevice': ('MassFlowController.m', 'Valve.m'),
         '@1D/Domain1D': ('1D/AxiStagnFlow.m', '1D/AxisymmetricFlow.m',
                          '1D/Inlet.m', '1D/Outlet.m', '1D/OutletRes.m',
                          '1D/Surface.m', '1D/SymmPlane.m'),
         '@1D/Stack': ('1D/FreeFlame.m', '1D/npflame_init.m'),
         '@Interface': ('importEdge.m', 'importInterface.m')}

for page in pages:
    ...
    for obj in page.objects:
        if obj.endswith('.m'):
            # do things with individual files
        else:
            # do things for a class
            for item in extra.get(obj, ()):
                # do things for extra files associated with this class
```

This gives you more control over the order that things are presented, and the `namedtuple` and auxiliary data structure help get rid of the confusing indexing.

File: doc/SConscript [hunk 5]
```diff
@@ -34,14 +34,170 @@ if localenv['sphinx_docs']:
...
+        # The base directory of the MATLAB toolbox relative to the sphinx build directory
+        base = '../interfaces/matlab/toolbox'
+        # Loop through the lists of lists
+        for lists in items[1:]:
+            # Loop through all the @folder names in the first nested list
+            for folder in lists[0]:
+                # Set the subheader based on the class name
+                doc += folder.split('@')[1] + '\n' + '-'*len(folder.split('@')[1]) + '\n\n'
+                # Alphabetically sort the list of files in the @folder
+                files = sorted(os.listdir(os.path.join(base,folder)))
```
Comment (speth) [inline doc/SConscript:168]: Anywhere in the `SCons*` files, you can use `pjoin` as a shorthand for `os.path.join`.

File: doc/sphinx/conf.py [hunk 6]
```diff
@@ -31,11 +31,19 @@
...
 # Add any Sphinx extension module names here, as strings. They can be extensions
 # coming with Sphinx (named 'sphinx.ext.*') or your custom ones.
-extensions = ['sphinx.ext.autodoc',
+
+# sphinxcontrib.matlab has been added to add the MATLAB domain for the 
+# documentation of the MATLAB functions. It is a new requirement to build the 
+# documentation in 2.2. It should be loaded before sphinx.ext.autodoc because 
+# loading it after gives errors when autodocumenting the Python interface.
+extensions = [
+              'sphinxcontrib.matlab',
```
Comment (speth) [inline doc/sphinx/conf.py:40]: This dependency needs to be added to `compiling.rst`. Alternatively, we could consider bundling this package (and maybe doxylink, too) in the `ext` directory.

**Comment by bryanwweber** [comment]
@speth The commits that you see here have already been squashed as much as I can make them. At this point, moving commits around causes the patches to no longer apply, and the rebase throws an error. I think it will be easier to cherry pick onto a new branch than to fix the conflicts from rebasing.

Thanks for the suggestion about `namedtuple`. That looks much neater and I'll definitely implement it. 

I'll add the new dependency to `compiling.rst`. It's available on the Cheese Factory (`pip install sphinxcontrib-matlabdomain` will get it), so I don't think we have to bundle it, and the same for `doxylink`. Sphinx is up to version 1.2 now, so would it make sense to switch to using Sphinx's `mathjax` extension instead of the custom one?

**Comment by speth** [comment]
@bryanwweber The current version of Sphinx isn't the limiting factor; it's the version of Sphinx available on platforms that we want to support. For example, Ubuntu 10.04 (still a supported LTS release) and RHEL6 only include Sphinx 0.6.x. I think we should at least wait until after Cantera 2.2 to increase the Sphinx version requirement. Although, I haven't tested the docs with older versions of Sphinx recently, so that may already be broken and there's no real change involved in officially requiring Sphinx 1.1+.

Pip only sometimes works on Windows, but since it's a pure Python module, installing it from the .zip shouldn't be that hard. It would be nice if this didn't introduce a new dependency for users who aren't building the Matlab toolbox. It's not a huge issue, since it only applies to the docs, which I assume not that many people build local copies of (though I could be wrong).

File: interfaces/matlab/toolbox/1D/@Stack/subsref.m [hunk 7]
```diff
@@ -1,5 +1,12 @@
 function b = subsref(s,index)
-% SUBSREF -
+% SUBSREF  Redefine subscripted references.
```
Comment (speth) [inline interfaces/matlab/toolbox/1D/@Stack/subsref.m:2]: Is there a way we can exclude docs for some of the internal functions like subsref and and the 'hndl' functions, which users should never need to interact with directly?

Comment (bryanwweber) [inline interfaces/matlab/toolbox/1D/@Stack/subsref.m:2]: We can add an if/else check in the SConscript, or we can add something to the `namedtuple` for the exclusions. I'll try to work something out.

File: interfaces/matlab/toolbox/Contents.m [hunk 8]
```diff
@@ -1,35 +1,45 @@
 % Cantera Toolbox.
-% Version 1.0   17-April-2002
```
Comment (speth) [inline interfaces/matlab/toolbox/Contents.m:2]: Do we need to provide `contents.m`? Matlab should autogenerate something that is guaranteed to be up to date. This file seems like the kind of thing that will always end up being out of date.

**Comment by bryanwweber** [comment]
@speth I'll switch the docs between `*molecularWeight` and `*molarMass`.

File: interfaces/matlab/toolbox/Contents.m (hunk 8 above)
Comment (bryanwweber) [inline interfaces/matlab/toolbox/Contents.m:2]: `Contents.m` could be generated by SConscript at build time. I don't know how to get MATLAB to autogenerate this file - the only way I know how is manually, as described [here](http://www.mathworks.com/help/matlab/matlab_prog/create-a-help-summary-contents-m.html) in the Matlab docs.

**Comment by speth** [comment]
@bryanwweber Aside from the comments I've left, I think these changes to the Matlab interface look good.

**Comment by bryanwweber** [comment]
@speth Thanks for your help Ray! I'm supposed to be writing my dissertation for the next few months, but I'll probably find a few minutes to work on this. Hopefully by the early summer (after I've graduated) I'll be able to come back to this full time if I haven't finished it up by then.

**Comment by bryanwweber** [comment]
I'm closing this PR and will commit a new set of changes to a new branch for review when they're finished.

**Comment by speth** [comment]
@bryanwweber According to http://www.mathworks.com/help/matlab/matlab_prog/create-a-help-summary-contents-m.html "Without a Contents.m file, the help and doc commands display a generated list of all program files in the folder." So if all we're providing is this info, we may as well not have this file in place.

**Comment by bryanwweber** [comment]
@speth The advantage of having the Contents.m file is that it can include general help links and copyright information, as well as providing the installed version number. It doesn't necessarily have to list all the functions in the toolbox, although that is MathWorks' default.

Unfortunately, to see that in Matlab, the user has to type `help toolbox`which is non-intuitive, and requires that the Cantera toolbox be first on their Matlab path.  Another option is to type `help Cantera\matlab\toolbox` (on Windows) which is slightly more intuitive and only requires that the `Cantera\matlab\toolbox` folder is somewhere on the path, which it has to be anyways for Cantera to work.

Either way, it seems like that would be a rarely used feature. One fix would be to rename the `toolbox` folder to something sensible like `Cantera` so the installation path would be `Cantera\matlab\Cantera` (again, on Windows). Then `help Cantera` or `help cantera` would show the contents of the proper `Contents.m`. Is there something special about the `toolbox` directory name in Matlab that requires its use in this case? I couldn't find anything on the internet, and just changing the folder name (and Matlab path) on my Windows install of 2.1 works in very limited testing...

**Comment by speth** [comment]
@bryanwweber Given the behavior of `help toolbox`, I think that implies that the directory name is intended to match the name of the toolbox, so changing the directory name to `cantera` is fine with me. Anything more complicated than `help cantera` seems like it would be unlikely to be used by anyone who actually needs the help.

If we want to have version information in the `Contents.m` file, then I think it should be generated from a template so that we can use `env['cantera_version']` to populate that. I'd prefer that any list of functions be auto-generated by SCons as well.

**Comment by bryanwweber** [comment]
The final commits were merged in by 02cd9a15692c30c408a6e050fb868fa52ab9be02

**Pull request closed without merging**